engine/execution/ingestion/stop/stop_control.go

```diff
@@ -0,0 +1,660 @@
+	"github.com/onflow/flow-go/model/flow"
+	"github.com/onflow/flow-go/module/component"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/state/protocol"
+	psEvents "github.com/onflow/flow-go/state/protocol/events"
+	"github.com/onflow/flow-go/storage"
+)
+
+// StopControl is a specialized component used by ingestion.Engine to encapsulate
+// control of stopping blocks execution.
+// It is intended to work tightly with the Engine, not as a general mechanism or interface.
+//
+// StopControl can stop execution or crash the node at a specific block height. The stop
+// height can be set manually or by the version beacon service event. This leads to some
+// edge cases that are handled by the StopControl:
+//
+//  1. stop is already set manually and is set again manually.
+//     This is considered as an attempt to move the stop height. The resulting stop
+//     height is the new one. Note, the new height can be either lower or higher than
+//     previous value.
+//  2. stop is already set manually and is set by the version beacon.
+//     The resulting stop height is the lower one.
+//  3. stop is already set by the version beacon and is set manually.
+//     The resulting stop height is the lower one.
+//  4. stop is already set by the version beacon and is set by the version beacon.
+//     This means version boundaries were edited. The resulting stop
+//     height is the new one.
+type StopControl struct {
+	// Stop control needs to consume BlockFinalized events.
+	// adding psEvents.Noop makes it a protocol.Consumer
+	psEvents.Noop
+	sync.RWMutex
+	component.Component
+
+	blockFinalizedChan chan *flow.Header
+
+	headers        StopControlHeaders
+	exeState       state.ReadOnlyExecutionState
+	versionBeacons storage.VersionBeacons
+
+	// stopped is true if node should no longer be executing blocks.
+	stopped bool
+	// stopBoundary is when the node should stop.
+	stopBoundary stopBoundary
+	// nodeVersion could be nil right now. See NewStopControl.
+	nodeVersion *semver.Version
+	// last seen version beacon, used to detect version beacon changes
+	versionBeacon *flow.SealedVersionBeacon
+	// if the node should crash on version boundary from a version beacon is reached
+	crashOnVersionBoundaryReached bool
+
+	log zerolog.Logger
+}
+
+var _ protocol.Consumer = (*StopControl)(nil)
+
+var NoStopHeight = uint64(math.MaxUint64)
+
+type StopParameters struct {
+	// desired StopBeforeHeight, the first value new version should be used,
+	// so this height WON'T be executed
+	StopBeforeHeight uint64
+
+	// if the node should crash or just pause after reaching StopBeforeHeight
+	ShouldCrash bool
+}
+
+func (p StopParameters) Set() bool {
+	return p.StopBeforeHeight != NoStopHeight
+}
+
+type stopBoundary struct {
+	StopParameters
+
+	// once the StopParameters are reached they cannot be changed
+	immutable bool
+
+	// This is the block ID of the block that should be executed last.
+	stopAfterExecuting flow.Identifier
```

Comment:
I feel like it's the last executed block before stop, right? Do we need thins only for logging purpose?

Reply:
we could be executing forks, se we need to know if we executed the block that was finalized.

That is why we store this value once we find the finalized block at the requested stop height.

Reply:
this goes away in a follow up PR https://github.com/onflow/flow-go/pull/4424